File: src/phase1/matrix/RotateMatrixAntiClockWIse.java
package phase1.matrix;

import static phase1.matrix.TransposeMatrixKt.transpose;

public class RotateMatrixAntiClockWIse {
    public static void main(String[] args) {
        int[][] input = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
        rotateMatrix(input);
        for (int i = 0; i < input.length; i++) {
            for (int j = 0; j < input[i].length; j++) {
                System.out.print(input[i][j]);
            }
            System.out.println();
        }
    }

    public static void rotateMatrix(int [][] matrix) {
        transpose(matrix);
        for (int i = 0; i <= matrix.length/2; i++) {
            for (int j = 0; j < matrix[i].length; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[matrix.length-1-i][j];
                matrix[matrix.length-1-i][j] = temp;
            }
        }
    }
}

File: src/phase1/matrix/RotateMatrixAnticlockWise.kt
package phase1.matrix

fun main() {
    rotateMatrixAntiClockWise(arrayOf(intArrayOf(1,2,3), intArrayOf(4,5,6), intArrayOf(7,8,9)))
}

fun rotateMatrixAntiClockWise(matrix: Array<IntArray>) {
    transposeMatrix(matrix)
    //interchange row first -><- last , second -><- second last
    for (i in 0..matrix.size / 2) {
        for (j in matrix[i].indices) {
            var temp = matrix[i][j]
            matrix[i][j] = matrix[matrix.size - 1 - i][j]
            matrix[matrix.size - 1 - i][j] = temp
        }
    }

    for (i in matrix.indices) {
        for (j in matrix[i].indices) {
            print(matrix[i][j])
        }
        println()
    }
}

fun transposeMatrix(matrix: Array<IntArray>) {
    for (i in matrix.indices) {
        for (j in i + 1 until matrix[i].size) {
            var temp = matrix[i][j]
            matrix[i][j] = matrix[j][i]
            matrix[j][i] = temp
        }
    }
}